src/Core/Appplication/Service/Telemetry/TelemetryEventsSynchronizer.php

```
@@ -0,0 +1,207 @@
+        $this->lockFactory = $lockFactory;
+        $this->isDebug = $isDebug;
+        $this->isTelemetryEnabled = $isTelemetryEnabled;
+    }
+
+    /**
+     * @param \SprykerSdk\SdkContracts\Entity\Telemetry\TelemetryEventInterface $telemetryEvent
+     *
+     * @return void
+     */
+    public function persist(TelemetryEventInterface $telemetryEvent): void
+    {
+        if (!$this->isTelemetryEnabled) {
+            return;
+        }
+
+        $this->telemetryEventRepository->save($telemetryEvent);
+    }
+
+    /**
+     * @return void
+     */
+    public function synchronize(): void
+    {
+        if (!$this->isTelemetryEnabled) {
+            return;
+        }
+
+        $this->cleanTelemetryEvents();
+
+        $lock = $this->lockFactory->createLock(static::LOCK_KEY, static::LOCK_TTL_SEC);
+
+        if (!$lock->acquire()) {
+            return;
+        }
+
+        try {
+            $this->synchronizeTelemetryEvents();
+        } finally {
+            $lock->release();
+        }
+    }
+
+    /**
+     * @return void
+     */
+    protected function cleanTelemetryEvents(): void
+    {
+        $this->telemetryEventRepository->removeAbandonedTelemetryEvents(
+            static::MAX_SYNCHRONIZATION_ATTEMPTS,
+            new DateInterval(sprintf('P%dD', static::MAX_EVENT_TTL_DAYS)),
+        );
+    }
+
+    /**
+     * @throws \SprykerSdk\Sdk\Infrastructure\Exception\TelemetryServerUnreachableException
+     *
+     * @return void
+     */
+    protected function synchronizeTelemetryEvents(): void
+    {
+        $syncTimestamp = (int)(new DateTimeImmutable())->format('Uu');
+
+        while (count($telemetryEvents = $this->getTelemetryEvents($syncTimestamp)) > 0) {
+            try {
+                $this->telemetryEventSender->send($telemetryEvents);
+                $this->telemetryEventRepository->removeBatch($telemetryEvents);
+            } catch (TelemetryServerUnreachableException $e) {
+                if ($this->isDebug) {
```

you need to save the exception to logs anyway. you ignore exception if it's not debugging. it's wrong